### DISCUSSION TITLE
More accurate Saving Session tarrifs

### DISCUSSION THREAD

**Discussion by Nilogax** [springfall2008/batpred#804] [Ideas]
At the moment predbat seems to assume that any export during a saving session will give the full benefit of the standard export rate plus saving session bonus.  I think Octopus actually work out the difference between the saving session export and your typical export during the time window, so the final benefit of the saving session may be substantially lower than the default assumption.

A simple solution would be to add a "saving session scale factor", perhaps as an expert mode setting. For example if the headline saving session bonus were £1.50 per kwh, a scale factor setting of 0.5 would mean that predbat assumes an actual bonus rate of £0.75 per kwh.  Users could then tweak this when a saving session is notified, depending on the time slot and their recent export pattern.   

I've tried using a workaround with the current system by changing the "octopoints per penny" tigure in apps.yaml. This then mpacts the p/kwh used by predbat but a separate setting as above would mean it could potentially be updated by an automation based on historic export data.  Given the time it takes for Octopus to actually come up with the results I'm assuming that their full algorithm is hugely complex and so not practical to implement!

**Comment by Nilogax** [comment]
The Octopus Saving Session algorithm doesn't actually seem that complex. It compares usage during the session with your baseline use in that time slot, and the baseline looks at the previous 10 weekdays or 4 weekend days which weren't affected by a saving session.

I'm no HA export but have bodged together a set of automations and a bit of Python scripting to retain the required data then calculate an adjusted "Octopoints per penny" figure assuming predbat will schedule a maximum export during the saving session. This should then reflect the likely financial benefit from exporting more than the baseline.

Of course I now need another couple of saving sessions to test it properly and compare my figures with the resulting Octopoints.

**Reply by iainfogg** [reply]
@Nilogax it may be that quite a few people might not have enough historic data in their HA system to do those calculations, I think a lot will only have 10 days as that's been the default in HA. 

So the calculation wouldn't be able to work properly without modifying the HA setup.

Sounds a neat idea though!

**Reply by Nilogax** [reply]
Yes I figured history might be an issue, especially if there were multiple sessions close to each other.  I chose to write the export data to a file each day at half hourly intervals from 4-7pm, then each night there's a script that either discards that day if there was a saving session or retains it and removes the previous 10 day old weekday or 4 day old weekend.  Anyhow, results are in from a 30 minute saving session on 14th Mar and although my own export history wasn't quite complete I estimated the net benefit as 450 octopoints and Octopus awarded 480, so I'll take that as a sign I'm on the right lines.  I doubt there will be many more saving sessions until the winter now, so further testing might need to wait.

**Reply by iainfogg** [reply]
It's confirmed there will be no more sessions this winter, so it'll need to wait until the end of the year now to test it further.

Predbat currently allows for this by simply adjusting the export rate to allow for the rate being offered by Octopus. It's not identical to your approach, but maybe is good enough.

See what @springfall2008 thinks...
